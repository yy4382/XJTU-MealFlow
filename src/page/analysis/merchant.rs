use std::collections::HashMap;

use ratatui::{
    Frame,
    layout::{Rect, Size},
    style::{Style, Stylize as _, palette::tailwind},
    symbols,
    text::Line,
    widgets::{Bar, BarChart, BarGroup, Block, Clear, Padding, Paragraph},
};
use tui_scrollview::{ScrollView, ScrollViewState, ScrollbarVisibility};

use crate::libs::transactions::Transaction;

#[derive(Debug, Default, Clone)]
pub(super) struct MerchantData {
    data: Vec<(String, f64)>,
    pub scroll_state: ScrollViewState,
}
impl MerchantData {
    pub fn new(data: &[Transaction]) -> Self {
        let hash_map = data.iter().fold(HashMap::new(), |mut acc, entry| {
            match acc.get(&entry.merchant) {
                Some(v) => acc.insert(&entry.merchant, *v + entry.amount),
                None => acc.insert(&entry.merchant, entry.amount),
            };
            acc
        });
        let mut entries: Vec<_> = hash_map.iter().collect();
        entries.sort_by(|a, b| a.1.total_cmp(b.1));
        MerchantData {
            data: entries.iter().map(|e| ((*e.0).to_string(), *e.1)).collect(),
            scroll_state: ScrollViewState::default(),
        }
    }
}

impl MerchantData {
    pub(super) fn render(
        &mut self,
        main_area: ratatui::prelude::Rect,
        frame: &mut Frame,
        color: tailwind::Palette,
    ) {
        let block = Block::bordered()
            .border_set(symbols::border::PROPORTIONAL_TALL)
            .border_style(color.c600)
            .padding(Padding::horizontal(1));

        if self.data.len() == 0 {
            frame.render_widget(
                Paragraph::new("No data available yet").block(block.clone()),
                main_area,
            );
            return;
        }

        let style = Style::default().fg(tailwind::BLUE.c300);
        let bars: Vec<Bar> = self
            .data
            .clone()
            .into_iter()
            .map(|(name, value)| {
                Bar::default()
                    .value(((value.abs() * 100.0).round() as u64) / 100)
                    .text_value(format!("{:.2}", value.abs()))
                    .label(Line::from(name))
                    .style(style)
                    .value_style(style.reversed())
            })
            .collect();
        let bar_chart = BarChart::default()
            .block(Block::default().padding(Padding::horizontal(1)))
            .data(BarGroup::default().bars(&bars))
            .bar_width(1)
            .bar_gap(1)
            .direction(ratatui::layout::Direction::Horizontal);

        // inset 1
        let chart_area = Rect {
            x: main_area.x + 1,
            y: main_area.y + 1,
            width: main_area.width - 2,
            height: main_area.height - 2,
        };
        frame.render_widget(block.clone(), main_area);
        frame.render_widget(Clear, chart_area);

        let chart_height = ((self.data.len()) * 2 - 1) as u16;

        let mut scroll_view = ScrollView::new(Size::new(chart_area.width, chart_height))
            .horizontal_scrollbar_visibility(ScrollbarVisibility::Never);

        scroll_view.render_widget(
            bar_chart,
            Rect::new(0, 0, chart_area.width - 1, chart_height),
        );

        frame.render_stateful_widget(scroll_view, chart_area, &mut self.scroll_state);
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;

    use super::*;

    #[test]
    fn test_empty_render() {
        let mut terminal = ratatui::Terminal::new(TestBackend::new(80, 20)).unwrap();
        let mut data = MerchantData::default();
        terminal
            .draw(|f| data.render(f.area(), f, tailwind::BLUE))
            .unwrap();
        assert_snapshot!(terminal.backend())
    }
}