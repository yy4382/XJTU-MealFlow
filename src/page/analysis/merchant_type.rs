use ratatui::{
    Frame,
    style::{Style, Stylize as _, palette::tailwind},
    symbols,
    text::Line,
    widgets::{Bar, BarChart, BarGroup, Block, Padding, Paragraph},
};

use crate::libs::transactions::Transaction;
use crate::utils::merchant_class::MerchantType; // 引入商家分类

#[derive(Debug, Default, Clone)]
pub(super) struct MerchantCategoryData {
    canteen_food: u32,
    canteen_drink: u32,
    supermarket: u32,
    bathhouse: u32,
    other: u32,
    unknown: u32,
}

impl MerchantCategoryData {
    pub(super) fn new(data: &[Transaction]) -> Self {
        data.iter().fold(Self::default(), |mut acc, entry| {
            let merchant_type = MerchantType::from_str(&entry.merchant);
            match merchant_type {
                MerchantType::CanteenFood => acc.canteen_food += 1,
                MerchantType::CanteenDrink => acc.canteen_drink += 1,
                MerchantType::Supermarket => acc.supermarket += 1,
                MerchantType::Bathhouse => acc.bathhouse += 1,
                MerchantType::Other => acc.other += 1,
                MerchantType::Unknown => acc.unknown += 1,
            }
            acc
        })
    }

    fn all_zero(&self) -> bool {
        self.canteen_food == 0
            && self.canteen_drink == 0
            && self.supermarket == 0
            && self.bathhouse == 0
            && self.other == 0
            && self.unknown == 0
    }
}

impl IntoIterator for &MerchantCategoryData {
    type Item = (&'static str, u32);
    type IntoIter = std::vec::IntoIter<Self::Item>;

    fn into_iter(self) -> Self::IntoIter {
        vec![
            ("食堂食物", self.canteen_food),
            ("食堂饮品", self.canteen_drink),
            ("超市", self.supermarket),
            ("浴室", self.bathhouse),
            ("其他", self.other),
            ("未知", self.unknown),
        ]
        .into_iter()
    }
}

impl MerchantCategoryData {
    pub(super) fn render(
        &self,
        area: ratatui::prelude::Rect,
        frame: &mut Frame,
        color: tailwind::Palette,
    ) {
        let block = Block::bordered()
            .border_set(symbols::border::PROPORTIONAL_TALL)
            .border_style(color.c600)
            .padding(Padding::horizontal(1));

        if self.all_zero() {
            frame.render_widget(Paragraph::new("暂无数据").block(block.clone()), area);
            return;
        }

        let style = Style::default().fg(color.c300);
        let bars: Vec<Bar> = self
            .into_iter()
            .map(|(name, value)| {
                Bar::default()
                    .value(u64::from(value))
                    .label(Line::from(name))
                    .style(style)
                    .value_style(style.reversed())
            })
            .collect();
        let bar_chart = BarChart::default()
            .block(block)
            .data(BarGroup::default().bars(&bars))
            .bar_width(1)
            .bar_gap(1)
            .direction(ratatui::layout::Direction::Horizontal);
        frame.render_widget(bar_chart, area);
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;

    use super::*;

    #[test]
    fn test_empty_render() {
        let mut terminal = ratatui::Terminal::new(TestBackend::new(80, 20)).unwrap();
        let data = MerchantCategoryData::default();
        terminal
            .draw(|f| data.render(f.area(), f, tailwind::BLUE))
            .unwrap();
        assert_snapshot!(terminal.backend())
    }
}
