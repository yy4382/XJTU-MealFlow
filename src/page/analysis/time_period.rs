use chrono::NaiveTime;
use ratatui::{
    Frame,
    style::{Style, Stylize as _, palette::tailwind},
    symbols,
    text::Line,
    widgets::{Bar, BarChart, BarGroup, Block, Padding, Paragraph},
};

use crate::libs::transactions::Transaction;

#[derive(Debug, Default, Clone)]
pub(super) struct TimePeriodData {
    /// 5am - 10:30am
    breakfast: u32,
    /// 10:30am - 1:30pm
    lunch: u32,
    /// 4:30pm - 7:30pm
    dinner: u32,
    /// other
    unknown: u32,
}

impl TimePeriodData {
    pub(super) fn new(data: &[Transaction]) -> Self {
        data.iter().fold(Self::default(), |mut acc, entry| {
            let time = entry.time.time();
            if Self::check_time_in(time, (5, 0), (10, 30)) {
                acc.breakfast += 1;
            } else if Self::check_time_in(time, (10, 30), (13, 30)) {
                acc.lunch += 1;
            } else if Self::check_time_in(time, (16, 30), (19, 30)) {
                acc.dinner += 1;
            } else {
                acc.unknown += 1;
            }
            acc
        })
    }
    fn check_time_in(time: NaiveTime, start: (u32, u32), end: (u32, u32)) -> bool {
        if time >= NaiveTime::from_hms_opt(start.0, start.1, 0).unwrap()
            && time < NaiveTime::from_hms_opt(end.0, end.1, 0).unwrap()
        {
            return true;
        }
        false
    }
    fn all_zero(&self) -> bool {
        self.breakfast == 0 && self.lunch == 0 && self.dinner == 0 && self.unknown == 0
    }
}
impl IntoIterator for &TimePeriodData {
    type Item = (&'static str, u32);
    type IntoIter = std::vec::IntoIter<Self::Item>;

    fn into_iter(self) -> Self::IntoIter {
        vec![
            ("Breakfast", self.breakfast),
            ("Lunch", self.lunch),
            ("Dinner", self.dinner),
            ("Other", self.unknown),
        ]
        .into_iter()
    }
}
impl TimePeriodData {
    pub(super) fn render(
        &self,
        area: ratatui::prelude::Rect,
        frame: &mut Frame,
        color: tailwind::Palette,
    ) {
        let block = Block::bordered()
            .border_set(symbols::border::PROPORTIONAL_TALL)
            .border_style(color.c600)
            .padding(Padding::horizontal(1));

        if self.all_zero() {
            frame.render_widget(
                Paragraph::new("No data available yet").block(block.clone()),
                area,
            );
            return;
        }

        let style = Style::default().fg(color.c300);
        let bars: Vec<Bar> = self
            .into_iter()
            .map(|(name, value)| {
                Bar::default()
                    .value(u64::from(value))
                    .label(Line::from(name))
                    .style(style)
                    .value_style(style.reversed())
            })
            .collect();
        let bar_chart = BarChart::default()
            .block(block)
            .data(BarGroup::default().bars(&bars))
            .bar_width(1)
            .bar_gap(1)
            .direction(ratatui::layout::Direction::Horizontal);
        frame.render_widget(bar_chart, area);
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;

    use super::*;

    #[test]
    fn test_empty_render() {
        let mut terminal = ratatui::Terminal::new(TestBackend::new(80, 20)).unwrap();
        let data = TimePeriodData::default();
        terminal
            .draw(|f| data.render(f.area(), f, tailwind::BLUE))
            .unwrap();
        assert_snapshot!(terminal.backend())
    }
}
