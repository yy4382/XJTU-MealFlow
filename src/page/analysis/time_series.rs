use std::collections::HashMap;

use chrono::Datelike;
use ratatui::{
    Frame,
    style::{Style, Stylize as _, palette::tailwind},
    symbols,
    text::Line,
    widgets::{Bar, BarChart, BarGroup, Block, Padding, Paragraph},
};
use tracing::info;

use crate::libs::transactions::Transaction;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
struct YearMonth {
    year: u16,
    month: u16,
}
impl YearMonth {
    fn new(year: u16, month: u16) -> Self {
        Self { year, month }
    }
    // fn relative_to(&self, other: &Self) -> f64 {
    //     let year_diff = self.year as f64 - other.year as f64;
    //     let month_diff = self.month as f64 - other.month as f64;
    //     year_diff * 12.0 + month_diff
    // }
    // fn relative_to_default(&self) -> f64 {
    //     let default = YearMonth::new(2015, 1);
    //     self.relative_to(&default)
    // }
    fn next(&self) -> Self {
        if self.month == 12 {
            YearMonth::new(self.year + 1, 1)
        } else {
            YearMonth::new(self.year, self.month + 1)
        }
    }
}
impl std::fmt::Display for YearMonth {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:02}-{:02}", self.year, self.month)
    }
}

#[derive(Debug, Default, Clone)]
pub(super) struct TimeSeriesData {
    data: Vec<(YearMonth, f64)>,
}

impl TimeSeriesData {
    pub(super) fn new(data: &[Transaction]) -> Self {
        let mut processed_data: Vec<(YearMonth, f64)> = data
            .iter()
            .fold(HashMap::new(), |mut acc, entry| {
                let year = entry.time.year() as u16;
                let month = entry.time.month() as u16;
                match acc.get(&(year, month)) {
                    Some(v) => acc.insert((year, month), *v + entry.amount.abs()),
                    None => acc.insert((year, month), entry.amount.abs()),
                };
                acc
            })
            .iter()
            .map(|((year, month), amount)| (YearMonth::new(*year, *month), *amount))
            .collect();

        processed_data.sort_by(|a, b| a.0.cmp(&b.0));

        let processed_data = processed_data
            .into_iter()
            .fold(Vec::new(), |mut acc, entry| {
                let mut last_ym = acc.last().unwrap_or(&entry).0;
                while last_ym.next() < entry.0 {
                    info!("Missing data for {:?}, {:?}", last_ym, entry.0);
                    acc.push((last_ym, 0.0));
                    last_ym = last_ym.next();
                }
                acc.push(entry);
                acc
            });

        info!("{:?}", processed_data);

        Self {
            data: processed_data,
        }
    }
    pub(super) fn render(
        &self,
        area: ratatui::prelude::Rect,
        frame: &mut Frame,
        color: tailwind::Palette,
    ) {
        let block = Block::bordered()
            .border_set(symbols::border::PROPORTIONAL_TALL)
            .border_style(color.c600)
            .padding(Padding::horizontal(1));

        if self.data.len() == 0 {
            frame.render_widget(
                Paragraph::new("No data available yet").block(block.clone()),
                area,
            );
            return;
        }

        let bars = (area.width as u64 - 6) / 8;

        let style = Style::default().fg(color.c300);
        let bars: Vec<Bar> = self.data[self.data.len().saturating_sub(bars as usize)..]
            .iter()
            .map(|(ym, value)| {
                Bar::default()
                    .value(value.round() as u64)
                    .label(Line::from(ym.to_string()))
                    .style(style)
                    .value_style(style.reversed())
            })
            .collect();

        let bar_chart = BarChart::default()
            .block(block)
            .data(BarGroup::default().bars(&bars))
            .bar_width(7)
            .bar_gap(1)
            .bar_style(style);

        frame.render_widget(bar_chart, area);
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;

    use super::*;

    #[test]
    fn test_empty_render() {
        let mut terminal = ratatui::Terminal::new(TestBackend::new(80, 20)).unwrap();
        let data = TimeSeriesData::default();
        terminal
            .draw(|f| data.render(f.area(), f, tailwind::BLUE))
            .unwrap();
        assert_snapshot!(terminal.backend())
    }
}
