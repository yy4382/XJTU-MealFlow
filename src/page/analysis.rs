use crossterm::event::KeyCode;
use merchant::MerchantData;
use merchant_type::MerchantCategoryData;
use ratatui::{
    layout::{Constraint, Layout},
    style::{Stylize, palette::tailwind},
    widgets::Tabs,
};
use strum::{Display, EnumIter, IntoEnumIterator};
use time_period::TimePeriodData;
use time_series::TimeSeriesData;

use crate::{
    actions::{ActionSender, LayerManageAction},
    app::layer_manager::EventHandlingStatus,
    libs::transactions::{Transaction, TransactionManager},
    tui::Event,
    utils::help_msg::{HelpEntry, HelpMsg},
};

use super::{EventLoopParticipant, Layer, WidgetExt};

mod merchant;
mod merchant_type;
mod time_period;
mod time_series;

pub(crate) struct Analysis {
    manager: crate::libs::transactions::TransactionManager,
    tx: ActionSender,

    analysis_type: AnalysisType,
    data: Vec<Transaction>,
}

#[derive(Display, EnumIter)]
enum AnalysisType {
    #[strum(to_string = "Time Period")]
    TimePeriod(TimePeriodData),
    #[strum(to_string = "Time Series")]
    TimeSeries(TimeSeriesData),
    #[strum(to_string = "Merchant")]
    Merchant(MerchantData),
    #[strum(to_string = "MerchantCategory")]
    MerchantCategory(MerchantCategoryData),
}

impl AnalysisType {
    fn next(&self, data: &[Transaction]) -> Self {
        match self {
            Self::TimePeriod(_) => Self::TimeSeries(TimeSeriesData::new(data)),
            Self::TimeSeries(_) => Self::Merchant(MerchantData::new(data)),
            Self::Merchant(_) => Self::MerchantCategory(MerchantCategoryData::new(data)),
            Self::MerchantCategory(_) => Self::TimePeriod(TimePeriodData::new(data)),
        }
    }
    fn previous(&self, data: &[Transaction]) -> Self {
        match self {
            Self::TimePeriod(_) => Self::MerchantCategory(MerchantCategoryData::new(data)),
            Self::MerchantCategory(_) => Self::Merchant(MerchantData::new(data)),
            Self::TimeSeries(_) => Self::TimePeriod(TimePeriodData::new(data)),
            Self::Merchant(_) => Self::TimeSeries(TimeSeriesData::new(data)),
        }
    }
    fn to_index(&self) -> usize {
        match self {
            AnalysisType::TimePeriod(_) => 0,
            AnalysisType::TimeSeries(_) => 1,
            AnalysisType::Merchant(_) => 2,
            AnalysisType::MerchantCategory(_) => 3,
        }
    }
    fn get_palette(&self) -> tailwind::Palette {
        match self {
            AnalysisType::TimePeriod(_) => tailwind::BLUE,
            AnalysisType::TimeSeries(_) => tailwind::GREEN,
            AnalysisType::Merchant(_) => tailwind::INDIGO,
            AnalysisType::MerchantCategory(_) => tailwind::YELLOW,
        }
    }
}

impl Analysis {
    pub fn new(tx: ActionSender, manager: TransactionManager) -> Self {
        let mut new = Self {
            manager,
            tx,
            analysis_type: AnalysisType::TimePeriod(Default::default()),
            data: vec![],
        };
        new.data = new
            .manager
            .fetch_all()
            .expect("Failed to load transactions");
        new.analysis_type = AnalysisType::TimePeriod(TimePeriodData::new(&new.data));
        new
    }
}

impl EventLoopParticipant for Analysis {
    fn handle_events(&mut self, event: &crate::tui::Event) -> EventHandlingStatus {
        let mut status = EventHandlingStatus::default();
        #[allow(clippy::single_match)]
        match event {
            Event::Key(key) => match key.code {
                KeyCode::Esc => {
                    self.tx.send(LayerManageAction::Pop);
                    status.consumed();
                }
                KeyCode::Char('h') | KeyCode::Left => {
                    self.analysis_type = self.analysis_type.previous(&self.data);
                    status.consumed();
                }
                KeyCode::Char('l') | KeyCode::Right => {
                    self.analysis_type = self.analysis_type.next(&self.data);
                    status.consumed();
                }
                KeyCode::Char('j') | KeyCode::Down => {
                    if let AnalysisType::Merchant(ref mut data) = self.analysis_type {
                        data.scroll_state.scroll_down();
                        status.consumed();
                    }
                }
                KeyCode::Char('k') | KeyCode::Up => {
                    if let AnalysisType::Merchant(ref mut data) = self.analysis_type {
                        data.scroll_state.scroll_up();
                        status.consumed();
                    }
                }
                _ => {}
            },
            _ => {}
        };
        status
    }
}

impl Layer for Analysis {}

impl WidgetExt for Analysis {
    fn render(&mut self, frame: &mut ratatui::Frame, area: ratatui::prelude::Rect) {
        let [header_area, main_area, help_area] = Layout::vertical([
            Constraint::Length(1),
            Constraint::Fill(1),
            Constraint::Length(3),
        ])
        .areas(area);

        let tabs = Tabs::new(AnalysisType::iter().map(|e| {
            format!(" {} ", e)
                .fg(tailwind::GRAY.c500)
                .bg(e.get_palette().c950)
        }))
        .select(self.analysis_type.to_index())
        .highlight_style((tailwind::GRAY.c200, self.analysis_type.get_palette().c600))
        .divider(" ")
        .padding("", "");

        frame.render_widget(tabs, header_area);

        let palette = self.analysis_type.get_palette();

        match &mut self.analysis_type {
            AnalysisType::TimePeriod(data) => data.render(main_area, frame, palette),
            AnalysisType::TimeSeries(data) => data.render(main_area, frame, palette),
            AnalysisType::Merchant(data) => data.render(main_area, frame, palette),
            AnalysisType::MerchantCategory(data) => data.render(main_area, frame, palette),
        };

        self.get_help_message().render(frame, help_area);
    }
}

impl Analysis {
    fn get_help_message(&self) -> HelpMsg {
        let mut help = HelpMsg::default();
        help.push(HelpEntry::new('h', "Last tab"));
        help.push(HelpEntry::new('l', "Next Tab"));
        help.push(HelpEntry::new(KeyCode::Esc, "Go back"));
        help
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::{actions::Action, libs::fetcher};
    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;
    use tokio::sync::mpsc::{self, UnboundedReceiver};

    fn get_test_objs() -> (UnboundedReceiver<Action>, Analysis) {
        let (tx, _rx) = mpsc::unbounded_channel();
        let manager = TransactionManager::new(None).unwrap();
        let data = fetcher::test_utils::get_mock_data(50);
        manager.insert(&data).unwrap();
        let page = Analysis::new(tx.clone().into(), manager);
        (_rx, page)
    }

    #[test]
    fn test_initial_state() {
        let (_, page) = get_test_objs();
        assert!(matches!(page.analysis_type, AnalysisType::TimePeriod(_)));
        assert!(!page.data.is_empty());
    }

    #[test]
    fn test_tab_navigation() {
        let (_, mut page) = get_test_objs();
        // Initial state should be TimePeriod
        assert!(matches!(page.analysis_type, AnalysisType::TimePeriod(_)));

        // Test moving to next tab
        page.handle_event_with_status_check(&'l'.into());
        assert!(matches!(page.analysis_type, AnalysisType::TimeSeries(_)));

        // Test moving back to TimePeriod
        page.handle_event_with_status_check(&'h'.into());
        assert!(matches!(page.analysis_type, AnalysisType::TimePeriod(_)));

        // Test wrapping around
        page.handle_event_with_status_check(&'h'.into());
        assert!(matches!(page.analysis_type, AnalysisType::Merchant(_)));
    }

    fn get_merchant_data(analysis_type: &AnalysisType) -> MerchantData {
        if let AnalysisType::Merchant(data) = analysis_type {
            return data.clone();
        } else {
            panic!("Should be merchant")
        }
    }

    #[test]
    fn test_scroll_merchant_data() {
        let (_, mut page) = get_test_objs();

        // First switch to Merchant tab
        page.handle_event_with_status_check(&'h'.into());

        let initial_offset = get_merchant_data(&page.analysis_type)
            .scroll_state
            .offset()
            .y;

        // Test scrolling down
        page.handle_event_with_status_check(&'j'.into());
        assert_eq!(
            get_merchant_data(&page.analysis_type)
                .scroll_state
                .offset()
                .y,
            initial_offset + 1
        );

        // Test scrolling up
        page.handle_event_with_status_check(&'k'.into());
        assert_eq!(
            get_merchant_data(&page.analysis_type)
                .scroll_state
                .offset()
                .y,
            initial_offset
        );
    }

    #[test]
    fn test_render_time_period() {
        let (_, mut page) = get_test_objs();
        let mut terminal = ratatui::Terminal::new(TestBackend::new(80, 20)).unwrap();

        terminal
            .draw(|f| {
                page.render(f, f.area());
            })
            .unwrap();

        assert_snapshot!(terminal.backend());
    }
    #[test]
    fn test_render_merchant() {
        let (_, mut page) = get_test_objs();
        let mut terminal = ratatui::Terminal::new(TestBackend::new(80, 20)).unwrap();
        // Switch to Merchant tab and render again
        page.handle_event_with_status_check(&'h'.into());

        terminal
            .draw(|f| {
                page.render(f, f.area());
            })
            .unwrap();

        assert_snapshot!(terminal.backend());

        let seq = "j".repeat(10);
        seq.chars().for_each(|c| {
            page.handle_event_with_status_check(&c.into());
        });
        terminal
            .draw(|f| {
                page.render(f, f.area());
            })
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
    #[test]
    fn test_render_time_series() {
        let (_, mut page) = get_test_objs();
        let mut terminal = ratatui::Terminal::new(TestBackend::new(80, 20)).unwrap();

        // Switch to TimeSeries tab and render again
        page.handle_event_with_status_check(&'l'.into());

        terminal
            .draw(|f| {
                page.render(f, f.area());
            })
            .unwrap();

        assert_snapshot!(terminal.backend());
    }
}
