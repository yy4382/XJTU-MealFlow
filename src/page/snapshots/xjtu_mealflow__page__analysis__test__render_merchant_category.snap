---
source: src/page/analysis.rs
expression: terminal.backend()
---
" Time Period   Time Series   Merchant   MerchantCategory                        "
"█▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀█"
"█ 食堂食物 19█████████████████████████████████████████████████████████████████ █" Hidden by multi-width symbols: [(3, " "), (5, " "), (7, " "), (9, " ")]
"█                                                                              █"
"█ 食堂饮品 11████████████████████████████████████                              █" Hidden by multi-width symbols: [(3, " "), (5, " "), (7, " "), (9, " ")]
"█                                                                              █"
"█ 超市     2██████                                                             █" Hidden by multi-width symbols: [(3, " "), (5, " ")]
"█                                                                              █"
"█ 浴室     14███████████████████████████████████████████████                   █" Hidden by multi-width symbols: [(3, " "), (5, " ")]
"█                                                                              █"
"█ 其他     0                                                                   █" Hidden by multi-width symbols: [(3, " "), (5, " ")]
"█                                                                              █"
"█ 未知     0                                                                   █" Hidden by multi-width symbols: [(3, " "), (5, " ")]
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄█"
"╭──────────────────────────────────────────────────────────────────────────────╮"
"│ Last tab: h | Next Tab: l | Go back: esc                                     │"
"╰──────────────────────────────────────────────────────────────────────────────╯"
